src/librustc_mir/transform/generator.rs

```diff
@@ -442,86 +442,74 @@ fn locals_live_across_suspend_points(
-        .into_engine(tcx, body_ref, def_id)
+        .into_results_cursor(body);
+
+    let mut init = MaybeInitializedLocals
+        .into_engine(tcx, body, def_id)
         .iterate_to_fixpoint()
-        .into_results_cursor(body_ref);
+        .into_results_cursor(body);
 
-    let mut storage_liveness_map = IndexVec::from_elem(None, body.basic_blocks());
-    let mut live_locals_at_suspension_points = Vec::new();
-    let mut live_locals_at_any_suspension_point = BitSet::new_empty(body.local_decls.len());
+    let mut live = MaybeLiveLocals
+        .into_engine(tcx, body, def_id)
+        .iterate_to_fixpoint()
+        .into_results_cursor(body);
 
-    for (block, data) in body.basic_blocks().iter_enumerated() {
-        if let TerminatorKind::Yield { .. } = data.terminator().kind {
-            let loc = Location { block, statement_index: data.statements.len() };
-
-            liveness.seek_to_block_end(block);
-            let mut live_locals = liveness.get().clone();
-
-            if !movable {
-                // The `liveness` variable contains the liveness of MIR locals ignoring borrows.
-                // This is correct for movable generators since borrows cannot live across
-                // suspension points. However for immovable generators we need to account for
-                // borrows, so we conseratively assume that all borrowed locals are live until
-                // we find a StorageDead statement referencing the locals.
-                // To do this we just union our `liveness` result with `borrowed_locals`, which
-                // contains all the locals which has been borrowed before this suspension point.
-                // If a borrow is converted to a raw reference, we must also assume that it lives
-                // forever. Note that the final liveness is still bounded by the storage liveness
-                // of the local, which happens using the `intersect` operation below.
-                borrowed_locals_cursor.seek_before_primary_effect(loc);
-                live_locals.union(borrowed_locals_cursor.get());
-            }
+    let mut borrowed = MaybeBorrowedLocals::all_borrows()
+        .into_engine(tcx, body, def_id)
+        .iterate_to_fixpoint()
+        .into_results_cursor(body);
 
-            // Store the storage liveness for later use so we can restore the state
-            // after a suspension point
-            storage_live.seek_before_primary_effect(loc);
-            storage_liveness_map[block] = Some(storage_live.get().clone());
+    // Liveness across yield points is determined by the following boolean equation, where `live`,
+    // `init` and `borrowed` come from dataflow and `movable` is a property of the generator.
+    // Movable generators do not allow borrows to live across yield points, so they don't need to
+    // store a local simply because it is borrowed.
+    //
+    //    live_across_yield := (live & init) | (!movable & borrowed)
```

Comment:
My assertion, which perhaps should have been stated explicitly as it relies on some MIR invariants, is that `live_across_yield` as currently formulated is always a strict subset of `storage_live`. This is because a local cannot be borrowed before it is `storage_live` and `borrowed` already invalidates borrows when it sees `StorageDead`. Likewise, a local cannot be initialized before it is marked `StorageLive` and it cannot be used (marking it live) after it is `StorageDead`.

Reply:
Ah yeah that makes sense. I'd forgotten that `borrowed` was cleared if it hit `StorageDead`. A comment to this effect (and possibly a `debug_assert`) would help clarify this!